geo/src/main/java/org/apache/carbondata/geo/QuadTreeCls.java

```diff
@@ -0,0 +1,904 @@
+    QuadRect outerRectangle = getOuterRectangle(vertexes);
+    if (this.root.getRect().outsideBox(outerRectangle)) {
+      LOGGER.warn("the query outer rect is bigger than the root node return");
+      return false;
+    }
+    return root.insert(vertexes);
+  }
+
+  /**
+   * Obtain the range of all nodes of tree species
+   * @return Scope List
+   */
+  public List<Long[]> getNodesData() {
+    List<Long[]> result = new ArrayList<>();
+    getNodeGridRange(root, result);
+    sortRange(result);
+    combineRange(result);
+    LOGGER.info("after query the range size is " + result.size());
+    return result;
+  }
+
+  /**
+   * Get the data range of a node
+   * @param node node
+   * @param result Return value
+   */
+  public void getNodeGridRange(QuadNode node, List<Long[]> result) {
+    if (node.getNodeStatus() == GridData.STATUS_ALL) {
+      Long[] range = node.getGrid().getHashIDRange();
+      result.add(range);
+    } else {
+      // The order of addition is, bottom left, top left, top right, bottom right
+      getSubChildGridRange(node, QuadNode.ChildEnum.BOTTOMLEFT, result);
+      getSubChildGridRange(node, QuadNode.ChildEnum.TOPLEFT, result);
+      getSubChildGridRange(node, QuadNode.ChildEnum.TOPRIGHT, result);
+      getSubChildGridRange(node, QuadNode.ChildEnum.BOTTOMRIGHT, result);
+    }
+  }
+
+  /**
+   * Get the range of a node's child nodes
+   * @param node tree node
+   * @param childType Child node type
+   * @param result  return value
+   */
+  public void getSubChildGridRange(QuadNode node, QuadNode.ChildEnum childType,
+                                   List<Long[]> result) {
+    QuadNode child = node.getChildren(childType);
+    if (child != null) {
+      getNodeGridRange(child, result);
+    }
+  }
+
+  /**
+   * Sorting and merging area results
+   * @param rangeList Region list
+   */
+  public void sortRange(List<Long[]> rangeList) {
+    // When sorting, you only need to sort the first node of long [],
+    // because you can ensure that the interval is not repeated
+    rangeList.sort(new Comparator<Long[]>() {
+      @Override
+      public int compare(Long[] x, Long[] y) {
+        return Long.compare(x[0], y[0]);
+      }
+    });
+  }
+
+  /**
+   * If the ordered data is merged, the number of data segments after merging may be reduced.
+   * @param rangeList Area list, sorted
+   */
+  public void combineRange(List<Long[]> rangeList) {
+    if (rangeList.size() > 1) {
+      for (int i = 0, j = i + 1; i < rangeList.size() - 1; i++, j++) {
+        long previousEnd = rangeList.get(i)[1];
+        long nextStart = rangeList.get(j)[0];
+        if (previousEnd + 1 == nextStart) {
+          rangeList.get(j)[0] = rangeList.get(i)[0];
+          rangeList.remove(i);
+        }
+      }
+    }
+  }
+
+  /**
+   * Get the circumscribed rectangle of a polygon
+   * @param polygon polygon
+   * @return Rectangular area
+   */
+  private QuadRect getOuterRectangle(List<double[]> polygon) {
```

Comment:
From the polygon coordinates, we are making MBR (minimum bounding rectangle) here. **This can include some extra points that are not there in actual polygon.** 
so quadtree may include those extra points in output ranges and carbon can give more rows than actual. 

Spark also need to apply filter (by directly calling st_contains()), only carbon filtering is not enough.
**This can give wrong data.** I don't see filtering code at spark in this PR.

please check. @VenuReddy2103 , @MarvinLitt 
cc: @ravipesala @kumarvishal09

Reply:
That's just the first step in the algorithm, to filter part of the data， 
In fact, the results are all accurate, no more and no less

Reply:
![image](https://user-images.githubusercontent.com/5889404/70708362-42c5cf00-1d15-11ea-9bd1-cfe8564f3c52.png)

Reply:
after MBR it includes red points also in the ranges. How can you say it is accurate ? spark also need to filter right ?

Reply:
// If the initial area is larger than the root node, exit directly
      QuadRect outerRectangle = getOuterRectangle(vertexes);
      if (this.root.getRect().outsideBox(outerRectangle)) {
        LOGGER.warn("the query outer rect is bigger than the root node return");
        return false;
      }
      return root.insert(vertexes);

getOuterRectangle is fist level check is a polygon is outsiebox with external rectangle then do not need to insert. you need to see the fun  root.insert(vertexes) that really insert the polygon to the map.

Reply:
we just want to insert the polygon to the reagon map, then we can get the hash id list